File: apps/desktop/src-tauri/src/command/bookmark.rs
use colette_api::{
    bookmark::{
        Bookmark, BookmarkCreate, BookmarkListQuery, BookmarkScrape, BookmarkScraped,
        BookmarkUpdate,
    },
    Paginated, Session,
};
use colette_core::bookmark::BookmarkService;
use tauri::State;
use uuid::Uuid;

#[tauri::command]
pub async fn list_bookmarks(
    service: State<'_, BookmarkService>,
    session: State<'_, Session>,
    query: BookmarkListQuery,
) -> Result<Paginated<Bookmark>, String> {
    let bookmarks = service
        .list_bookmarks(query.into(), session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(bookmarks.into())
}

#[tauri::command]
pub async fn create_bookmark(
    service: State<'_, BookmarkService>,
    session: State<'_, Session>,
    data: BookmarkCreate,
) -> Result<Bookmark, String> {
    let bookmark = service
        .create_bookmark(data.into(), session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(bookmark.into())
}

#[tauri::command]
pub async fn get_bookmark(
    service: State<'_, BookmarkService>,
    session: State<'_, Session>,
    id: Uuid,
) -> Result<Bookmark, String> {
    let bookmark = service
        .get_bookmark(id, session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(bookmark.into())
}

#[tauri::command]
pub async fn update_bookmark(
    service: State<'_, BookmarkService>,
    session: State<'_, Session>,
    id: Uuid,
    data: BookmarkUpdate,
) -> Result<Bookmark, String> {
    let bookmark = service
        .update_bookmark(id, data.into(), session.profile_id)
        .await
        .map_err(|e| e.to_string())?;

    Ok(bookmark.into())
}

#[tauri::command]
pub async fn delete_bookmark(
    service: State<'_, BookmarkService>,
    session: State<'_, Session>,
    id: Uuid,
) -> Result<(), String> {
    service
        .delete_bookmark(id, session.profile_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn scrape_bookmark(
    service: State<'_, BookmarkService>,
    data: BookmarkScrape,
) -> Result<BookmarkScraped, String> {
    let scraped = service
        .scrape_bookmark(data.into())
        .await
        .map_err(|e| e.to_string())?;

    Ok(scraped.into())
}

File: apps/desktop/src/commands/bookmark.ts
import {
  Bookmark,
  type BookmarkAPI,
  BookmarkCreate,
  BookmarkList,
  BookmarkListQuery,
  BookmarkScrape,
  BookmarkScraped,
  BookmarkUpdate,
} from '@colette/core'
import { invoke } from '@tauri-apps/api/core'

export class BookmarkCommands implements BookmarkAPI {
  list(query: BookmarkListQuery): Promise<BookmarkList> {
    return invoke('list_bookmarks', {
      query: BookmarkListQuery.parse(query),
    }).then(BookmarkList.parse)
  }

  get(id: string): Promise<Bookmark> {
    return invoke('get_bookmark', { id }).then(Bookmark.parse)
  }

  create(data: BookmarkCreate): Promise<Bookmark> {
    return invoke('create_bookmark', { data: BookmarkCreate.parse(data) }).then(
      Bookmark.parse,
    )
  }

  update(id: string, data: BookmarkUpdate): Promise<Bookmark> {
    return invoke('update_bookmark', {
      id,
      data: BookmarkUpdate.parse(data),
    }).then(Bookmark.parse)
  }

  delete(id: string): Promise<void> {
    return invoke('delete_bookmark', { id }).then()
  }

  scrape(data: BookmarkScrape): Promise<BookmarkScraped> {
    return invoke('scrape_bookmark', { data: BookmarkScrape.parse(data) }).then(
      BookmarkScraped.parse,
    )
  }
}
